lib/sequelize.js

```diff
@@ -974,41 +975,61 @@ class Sequelize {
     // testhint argsConform.end
 
     const transaction = new Transaction(this, options);
-    const ns = Sequelize.cls;
-
-    if (autoCallback) {
-      let transactionResolver = (resolve, reject) => {
-        transaction.prepareEnvironment().then(() => {
-          if (ns) {
-            autoCallback = ns.bind(autoCallback);
-          }
-
-          const returnValue = autoCallback(transaction);
-          if (!returnValue || !returnValue.then) throw new Error('You need to return a promise chain/thenable to the sequelize.transaction() callback');
-
-          return returnValue.then(result => transaction.commit()).then(() => {
-            resolve(returnValue);
-          });
-        }).catch(err => {
-          // If the transaction has already finished (commit, rollback, etc), reject with the original error
-          if (transaction.finished) {
-            reject(err);
-          } else {
-            return transaction.rollback().finally(() => {
-              reject(err);
-            });
-          }
-        });
-      };
-
-      if (ns) {
-        transactionResolver = ns.bind(transactionResolver, ns.createContext());
-      }
+    const ns = Sequelize._cls;
+
+    if (!autoCallback) return transaction.prepareEnvironment().return(transaction);
+
+    // autoCallback provided
+    const wrapper = ns ? function(fn) {
+      var promise;
+      ns.run(() => promise = fn());
+      return promise;
+    } : function(fn) { return fn(); };
+
+    return wrapper(() =>
+      transaction.prepareEnvironment().then(() =>
+        autoCallback(transaction)
```

Comment:
Hmm yeah. First time I've used ES6 short function notation. I normally write with backward compat to node 0.12 in mind.

It's tricky because since none of that block has any returns at all, taking that logic to its furthest end, you'd end up with the entire block being a single long line of code. But I see your point.

I've added some more return statements. They're unnecessary, but I think it preserves readability while avoiding the problem you raise.